extensions/storages/source/providers/S3.ts

```diff
@@ -0,0 +1,101 @@
+import { type Readable } from 'node:stream'
+import { join } from 'node:path'
+import {
+  S3Client,
+  GetObjectCommand,
+  PutObjectCommand,
+  CopyObjectCommand,
+  ListObjectsV2Command,
+  DeleteObjectsCommand
+} from '@aws-sdk/client-s3'
+import { type Provider } from '../Provider'
+
+const {
+  S3_REGION = '',
+  S3_ENDPOINT = '',
+  S3_ACCESS_KEY = '',
+  S3_SECRET_ACCESS_KEY = '',
+  S3_BUCKET = ''
+} = process.env
+
+export class S3 implements Provider {
+  protected readonly path: string
+  protected readonly bucket = S3_BUCKET
+  protected readonly client: S3Client
+
+  public constructor (url: URL) {
+    if (url.host !== '')
```

Comment:
Pay attention to [url format](https://github.com/toa-io/toa/tree/feat/storages/extensions/storages#amazon-s3) for S3 provider.

Reply:
To configure the AWS SDK client for S3, I need to obtain the Endpoint and Bucket information. Currently, I assume that I can use a URL in the format "s3:///some_folder1/some_folder2". I will retrieve the Endpoint and Bucket for client configuration from environment variables.

However, I can change the URL format if needed. For instance, something like "s3:///s3.endpoint.cloud/some_bucket/some_folder" could be used. In this case, I would extract the Endpoint from the URL's host (s3.endpoint.cloud) and the Bucket from the first part of the pathname (some_bucket).